### ISSUE TITLE
DataFrame.apply unintuitively changes int to float because of another column

### ISSUE THREAD

**Issue by sergeny** [pandas-dev/pandas#23230]
#### Code Sample, a copy-pastable example if possible

```python
>>> import pandas as pd

# This works.

>>> pd.DataFrame({'a': [1, 2, 3], 'b': ['', '', '']}).apply(lambda row: row['a'], axis=1)
0    1
1    2
2    3
dtype: int64
    
# Here we also expect 1, 2, 3, as above.

>>> pd.DataFrame({'a': [1, 2, 3], 'b': [0., 0., 0.]}).apply(lambda row: row['a'], axis=1)
0    1.0
1    2.0
2    3.0

# Why floats?!?!?!?!?!

# It's an integer column:

>>> pd.DataFrame({'a': [1, 2, 3], 'b': [0., 0., 0.]})['a'].dtype
dtype('int64')
```
#### Problem description

The column 'a' above has type integer, not float. The `apply` function should not change the type just because the dataframe has another, unrelated float column.

I understand, why it happens: it detects the most suitable type for a Series. I still consider it unintuitive that I select a group of columns to apply some function to them that only works on ints, not on floats, and suddenly I remove one unrelated column and get an exception, because now I only have numeric columns, and all ints became floats.

#### Expected Output
```
0    1
1    2
2    3
dtype: int64
```

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.6.final.0
python-bits: 64
OS: Darwin
OS-release: 17.4.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.23.4
pytest: None
pip: 10.0.1
setuptools: 40.0.0
Cython: None
numpy: 1.15.0
scipy: 1.1.0
pyarrow: 0.9.0
xarray: None
IPython: 6.5.0
sphinx: None
patsy: 0.5.0
dateutil: 2.7.3
pytz: 2018.5
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.2.3
openpyxl: None
xlrd: 1.1.0
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 1.0.1
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
</details>

**Comment by TomAugspurger** [comment]
As the docstring says,

> Objects passed to the function are Series objects whose index is
either the DataFrame's index (``axis=0``) or the DataFrame's columns
(``axis=1``). 

so your lambda is getting a row like

```
In [10]: df.loc[0]
Out[10]:
a    1.0
b    0.0
Name: 0, dtype: float64
```

which has float dtype (float is the common dtype of int and float).

**Comment by jhylands** [comment]
Stack overflow solution:

https://stackoverflow.com/questions/55574348/dataframe-apply-unintuitively-changes-int-to-float-breaking-an-index-loopup

**Comment by Ricyteach** [comment]
I agree this is very unintuitive behavior and it caused me some difficulty this weekend. Here is my SO question to give context:

https://stackoverflow.com/questions/59240658/integer-format-specification-d-produces-valueerror-when-applied-row-by-row-to/59240771

I suggest it would be helpful to add some specific discussion about this behavior to the `DataFrame.apply()` docs page, since that is the place where most people will be looking into doing row-wise operations (with axis=1). And also, perhaps, `DataFrame.transpose()`.

**Comment by smcinerney** [comment]
This issue was closed but still in 1.0.x seems like a valid docbug on [`DataFrame.apply`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.apply.html), could you reopen as 'DOCBUG: ... for 1.0.4,' or else let's incorporate this docbug in related docbug #30191?

Here is an example dataframe which exercises upcasting of dtypes in broadcasting with `apply(..., axis=1)`:

```
df = pd.DataFrame({'i': [5, 6, 7], 'j': [3, 4, 2],
                   'f': [0.2, 1.4, np.nan], 's': ['x','y','z'],
                   'l': [[2,3], [4,6], [5,1]], 'b': [True, False, True],
                   'ct': pd.Categorical([7,8,9]),
                   'dt': pd.to_datetime(['2019/01/31', '2019/02/01', '2019/06/14']) })
```

Below is some suggested documentation for [`DataFrame.apply` manpage](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.apply.html).
Really there isn't any central section in the doc which we can reference and say "Beware that <Section> on coercing applies when a series has mixed types"; one sentence below is adapted from [basics doc](https://pandas.pydata.org/pandas-docs/stable/getting_started/basics.html#basics-dtypes).

> - When axis=1, each row is passed as a Series.
> - Hence when a row has mixed dtypes, beware that the dtype of the Series passed to `apply`
will be upcast/chosen to accommodate all of the data types in the entire row ('object' is the most general, and 'float' for all numeric dtypes).
> - However pandas will attempt to downcast the result to accommodate the original dtypes where possible:

```
>>> df.apply(lambda row: row['i'] + row['j'], axis=1)
0     8
1    10
2     9
dtype: int64
```
